### PR TITLE
Feature/mvp algo

### PR DESCRIPTION
**Pull request by bat3**
Feature to upload python algo and execute it on a selected series.

### CODE REVIEW

File: business/IA/basic_python.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import SimpleITK as sitk
```
Comment (bat3) [inline business/IA/basic_python.py]: To delete

File: business/MetadataDatabase/Controllers/AlgoController.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        public ActionResult<IEnumerable<AlgoDto>> Get()
+        {
+            return this.Ok(this.algoService.GetAll());
+        }
+
+        // GET: api/Algo/5
+        [HttpGet("{id}", Name = "GetAlgo")]
+        public ActionResult<AlgoDto> Get(string id)
+        {
+            return this.algoService.Get(id);
```
Comment (ptitdam2001) [inline business/MetadataDatabase/Controllers/AlgoController.cs]: Check if not exists

File: business/MetadataDatabase/Controllers/AlgoController.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        /// Creates an Algo.
+        /// </summary>
+        [HttpPost]
+        public ActionResult<LabelDto> Post([FromBody] AlgoDto algo)
+        {
+            var algoWithId = this.algoService.Create(algo);
+            try
+            {
+                Directory.CreateDirectory("workspace/algos/");
+                System.IO.File.WriteAllText(("workspace/algos/" + algo.MainFile), algo.ContentFile);
```
Comment (ptitdam2001) [inline business/MetadataDatabase/Controllers/AlgoController.cs:86]: Maybe I have put the file content into one collection.
In the future, we will be able to add a "code editor" in the frontend to edit and download the algorithm in python file.

Comment (ptitdam2001) [reply business/MetadataDatabase/Controllers/AlgoController.cs:86]: I don't know if it is possible to exec an eval of string in Python

File: business/MetadataDatabase/Controllers/Labels/LabelController.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        public ActionResult<IEnumerable<LabelDto>> GetAllLabels()
+        {
+            return this.Ok(this.labelService.GetAll());
+        }
+
+        // GET: api/Label/5
+        [HttpGet("{id}", Name = "GetLabel")]
+        public ActionResult<LabelDto> GetLabel(string id)
+        {
+            return this.labelService.Get(id);
```
Comment (ptitdam2001) [inline business/MetadataDatabase/Controllers/Labels/LabelController.cs]: Check if it exists

File: business/MetadataDatabase/Controllers/Labels/LabelController.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        // POST: api/Label
+        /// <summary>
+        /// Creates a Label.
+        /// </summary>
+        [HttpPost]
+        public ActionResult<LabelDto> Post([FromBody] LabelDto label)
```
Comment (ptitdam2001) [inline business/MetadataDatabase/Controllers/Labels/LabelController.cs]: If I were you I verify if the label id is null, I am not sure that to have one void string as ID generates one database Id

File: business/MetadataDatabase/Services/AlgoService.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            File.Delete(zipFilzName);
+
+            // Execute algo
+            algoExeInfo.Folder = relatifPathOfSeriesDirectory;
+            Console.WriteLine(algoExeInfo.Folder);
+            var jsonString = JsonSerializer.Serialize(algoExeInfo);
+            var content = new StringContent(jsonString, Encoding.UTF8, "application/json");
+            var task = this.client.PostAsync($"{this.settings.Protocol}://{this.settings.Host}:{this.settings.Port}/executeAlgo", content);
+            var res = task.Result.Content.ReadAsStringAsync();
+            return res.Result;
```
Comment (ptitdam2001) [inline business/MetadataDatabase/Services/AlgoService.cs:75]: You should delete unzipped files at the end of execution

File: business/MetadataDatabase/Services/Labels/LabelService.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+        public IEnumerable<LabelDto> GetAll()
+        {
+            return this.labelRepository.GetAll().ToDto();
+        }
+
+        public void Update(string id, LabelDto objectToUpdate)
+        {
+            objectToUpdate.Id = id;
+            this.labelRepository.Update(objectToUpdate.ToModel());
```
Comment (ptitdam2001) [inline business/MetadataDatabase/Services/Labels/LabelService.cs]: Should you call IsValid method before update?

File: business/MetadataDatabase/Services/Labels/LabelService.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        private readonly ILabelRepository labelRepository;
+
+        public LabelService(ILabelRepository labelRepository)
+        {
+            this.labelRepository = labelRepository;
+        }
+
+        public LabelDto Create(LabelDto objectToCreate)
+        {
+            return this.labelRepository.Create(objectToCreate.ToModel()).ToDto();
```
Comment (ptitdam2001) [inline business/MetadataDatabase/Services/Labels/LabelService.cs]: Should you call IsValid method before create?

File: client/web/src/App.tsx [hunk 9]
```diff
@@ -46,6 +47,9 @@ const App: React.FC = () => {
             <ConnectedRoute path="/series" exact>
               <Series />
             </ConnectedRoute>
+            <ConnectedRoute path="/algorithms" exact>
+              <Algorithms />
```
Comment (ptitdam2001) [inline client/web/src/App.tsx:51]: Ah shit!! we will have a conflict with my branch

File: client/web/src/components/algo/AlgoResultDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  return (
+    <div>
+    {
+    shouldBeOpen() 
+    ? 
+      <Dialog open={props.open} onClose={handleClose} aria-labelledby="form-dialog-title">
+        <DialogTitle id="form-dialog-title">
+        Algo result
```
Comment (ptitdam2001) [inline client/web/src/components/algo/AlgoResultDialog.tsx]: just introduce the T function for translation

File: client/web/src/components/labels/LabelAdding.tsx [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+    itemLabelizable.labels.push(newSelectedLabel);
+    dispatch({ type: actionTypes.UPDATE_SERIES, series: itemLabelizable});
+    if(props.onCreate !== undefined) props.onCreate();
+  }
+
+  const checkInputs = () => {
+    if(selectedLabelValue !==  '') 
+    {
+      setAreInputsValid(true)
```
Comment (ptitdam2001) [inline client/web/src/components/labels/LabelAdding.tsx]: simplify with `setAreInputsValid(selectedLabelValue !== '')`

File: client/web/src/components/labels/LabelAdding.tsx [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    } else
+    {
+      setAreInputsValid(false)
+    }
+  }
+
+  useEffect(checkInputs, [selectedLabelValue]);
+
+  return (
+    <form>
```
Comment (ptitdam2001) [inline client/web/src/components/labels/LabelAdding.tsx]: you can use formik, you will have less code

File: client/web/src/components/labels/LabelCreation.tsx [hunk 13]
```diff
@@ -0,0 +1,168 @@
+import React, { useState, useEffect } from 'react'
```
Comment (ptitdam2001) [inline client/web/src/components/labels/LabelCreation.tsx]: There will be some conflicts with my branch!

File: client/web/src/components/labels/LabelManagementDialog.tsx [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+enum Mode {
+  Creation = 1,
+  Adding,
+} 
+
+export const LabelManagementDialog: React.FC<LabelManagementDialogProps> = (props: LabelManagementDialogProps) => {
+  const { t } = useTranslation()
+  const propIsOpen = props.open;
```
Comment (ptitdam2001) [inline client/web/src/components/labels/LabelManagementDialog.tsx]: you can use detructuring, for here, you can change by: 

```typescript
const { open: propIsOpen, itemLabelizable, onClose } = props
```
After that you don't need to use the variable props

File: client/web/src/components/labels/LabelManagementDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+            mode === Mode.Adding
+            ? <LabelAdding itemLabelizable={itemLabelizable} onCancel={handleClose} onCreate={handleClose}/>
+            : <LabelCreation onCreate={handleLabelAdding} onCancel={handleLabelAdding}/>
+          }
+        </DialogContent>
+        <DialogActions>
+          {
+            mode === Mode.Adding
+            ? <Button onClick={handleLabelCreation} color="primary">
+                Create a new label
```
Comment (ptitdam2001) [inline client/web/src/components/labels/LabelManagementDialog.tsx]: add t('') function ;)

**Review by ptitdam2001 — CHANGES_REQUESTED** [review]

File: business/MetadataDatabase/Controllers/Labels/LabelController.cs (hunk 5 above)
Comment (bat3) [reply business/MetadataDatabase/Controllers/Labels/LabelController.cs]: id is auto generated by mongo

File: business/MetadataDatabase/Services/AlgoService.cs [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            this.settings = settings.Value;
+            this.algoRepository = algoRepository;
+            this.pacsService = pacsService;
+            this.client = new HttpClient();
+        }
+
+        public string Execute(AlgoExeInfoDto algoExeInfo)
+        {
+            Directory.CreateDirectory("workspace/data/");
+            var zipFilzName = this.pacsService.DownloadSeries(algoExeInfo.SeriesInstanceUID);
```
Comment (cquennouelle) [inline business/MetadataDatabase/Services/AlgoService.cs]: zipFileName

File: business/IA/basic_python.py (hunk 1 above)
Comment (cquennouelle) [reply business/IA/basic_python.py]: Indeed file in double

File: business/MetadataDatabase/Services/Labels/LabelService.cs (hunk 7 above)
Comment (bat3) [reply business/MetadataDatabase/Services/Labels/LabelService.cs]: Done after merge develop

File: business/MetadataDatabase/Services/Labels/LabelService.cs (hunk 8 above)
Comment (bat3) [reply business/MetadataDatabase/Services/Labels/LabelService.cs]: Done after merge develop

File: business/MetadataDatabase/Controllers/Labels/LabelController.cs (hunk 5 above)
Comment (bat3) [reply business/MetadataDatabase/Controllers/Labels/LabelController.cs]: Tested with swagger there is no problems

File: business/MetadataDatabase/Controllers/AlgoController.cs (hunk 3 above)
Comment (bat3) [reply business/MetadataDatabase/Controllers/AlgoController.cs:86]: Done

File: client/web/src/components/labels/LabelAdding.tsx (hunk 12 above)
Comment (bat3) [reply client/web/src/components/labels/LabelAdding.tsx]: Formik is not working with my changing form...

**Review by ptitdam2001 — APPROVED** [review]

**Pull request merged**